docs/rfcs/033-merged-compute-image.md

```diff
@@ -0,0 +1,54 @@
+## Merged compute image
+https://github.com/neondatabase/neon/issues/6685
+
+### Motivation:
+It's hard to manage compute pools for 3 Postgres versions.  
+(we have a compute image for each version of Postgres (currently, it's 3 for neonVM and 3 for k8s pods; eventually, we will have only neonVMs)).
+We can try putting all Postgres versions into a single image, which should dramatically improve pool usage.
```

Comment:
It was initially written before we switched to dynamic pools, I would be interested to get the @mtyazici' opinion/estimations regarding it. 
I guess that managing a single pool is easier than managing three pools, and we can predict it better.

Reply:
> I guess managing a single pool is easier than managing three pools, and we can predict it better.

Before migrating to VMs we were managing tens of pools for a each region, it was quite complicated. Now, it's a lot better. Merging 3 pg versions into single image will probably make it easier to manage and estimate. Though I still have some concerns: 

- Adding 3 timeseries into a single one and forecasting it could actually be harder than forecasting 3 separate timeseries and then adding the forecasts. Either way, we should be able to implement either solution without too much of a problem, though first we need to check if there is a significant difference between them.
- Merging 3 pools into a single one will require some coordination and probably an intermediate/hacky state for cplane pool fill job and forecasting service. Don't really know what the best approach would be but wouldn't expect a significant blocker.

Reply:
I have [this document](https://www.notion.so/neondatabase/Compute-Pools-Autoscaling-3c34859083284154bc94d05b50233789). You can read my last comment at the top just before the Lassi's final comment, the second half is a about pool size forecasting. It's very similar to what I did in the end.

Though, I should clean up the document to describe the final result.

Reply:
Hm, I'd say that initial motivation

> It's hard to manage compute pools for 3 Postgres versions.
> We can try putting all Postgres versions into a single image, which should dramatically improve pool usage.

is no longer super relevant.

For the last month+ compute pools are in the free flight and we have [pretty cool hit rate ~99%](https://neonprod.grafana.net/d/ade9uqqklodmob/neon-control-plane-slos?orgId=1&var-datasource=ZNX49CDVz&var-service=control-plane&var-slo=pool_hit_rate_over_p99&var-region=All&var-sli_window=5m&from=now-30d&to=now). And managing 3 pools per region is not like managing tens of them, so 1 vs. 3 is still comparable.

I think there is one case, when merging could be beneficial in this sense -- when there is long irregular tale of older version start. In this case we will fail to predict demand and users will more likely to miss pools. At the same time, having a small static pool for old PGs would work too. Yet, so far [even PG 14 has small regular rate of starts](https://neonprod.grafana.net/d/a09466b6-1dc5-496a-b7f4-bd70a14aa28d/neon-compute-pools?orgId=1&var-env=prod&var-env_dev_to_staging=prod&var-percentile=0.9&var-source=ZNX49CDVz&var-compute_k8s_source=ZNX49CDVz&var-logs_source=grafanacloud-logs&var-service=neon-control-plane-api&var-cu=All&var-provisioner=All&var-region_id=All&var-pg_version=14&var-predictor=size_limiter&var-bucket_duration=30s&var-az=All&from=now-24h&to=now)

I don't write this to say that we don't need merging at all, maybe there are other benefits, which we can list here